### PR TITLE
Store backups in 1Password vault

### PR DESCRIPTION
**Pull request by dalenewby**
Dump variables for all workspaces, tar them up, and store in a 1Password vault using a Docker image.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -36,3 +36,4 @@ inc/
 
 # tfc-dump creates .json files that don't belong in version control
 *.json
+local.env
```
Comment (Baggerone) [inline .gitignore]: Maybe just `*.env` ?

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,21 @@
+FROM alpine:3.18
+
+ENV TFC_OPS_VER="3.5.1"
+
```
Comment (Baggerone) [inline Dockerfile:5]: (This is low importance.) I wonder if you could also have
`ENV TFC_OPS_GZ="tfc-ops_${TFC_OPS_VER}_Linux_x86_64.tar.gz"`
and use that it lines 11-13

**Review by Baggerone — APPROVED** [review]

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,22 @@
+FROM alpine:3.18
```
Comment (briskt) [inline Dockerfile:1]: While it's good to specify a version, does it need to be this specific? Maybe just `FROM alpine:3`. Not saying it needs to be looser, just asking.

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+ENV TFC_OPS_VER="3.5.1"
+ENV TFC_OPS_DISTRIB="tfc-ops_${TFC_OPS_VER}_Linux_x86_64.tar.gz"
+
+COPY ./dump-and-save.sh  /usr/local/bin/dump-and-save.sh
+COPY ./tfc-dump.pl       /usr/local/bin/tfc-dump.pl
+
+# Install tfc-ops, 1Password CLI, perl, and jq
+RUN chmod +x /usr/local/bin/dump-and-save.sh /usr/local/bin/tfc-dump.pl \
```
Comment (briskt) [inline Dockerfile]: git should retain the execute bit, so the `chmod` should not be necessary

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+ && cd /tmp \
+ && wget https://github.com/silinternational/tfc-ops/releases/download/v${TFC_OPS_VER}/${TFC_OPS_DISTRIB} \
+ && tar zxf ${TFC_OPS_DISTRIB} \
+ && rm LICENSE README.md ${TFC_OPS_DISTRIB} \
+ && mv tfc-ops /usr/local/bin \
+ && chmod +x /usr/local/bin/tfc-ops \
```
Comment (briskt) [inline Dockerfile:15]: likewise here, zip should retain the execute bit so the `chmod` isn't strictly necessary I believe

File: README.md [hunk 6]
```diff
@@ -37,3 +37,18 @@ Spaces in the variable set name are replaced with hyphens (`-`).
...
+2. Obtain a Terraform Cloud access token. Go to https://app.terraform.io/app/settings/tokens to create an API token.
+3. Add the access token value to `local.env`.
+4. If needed, create a 1Password vault to store the backup files in.
+5. Add the 1Password vault name and item name to `local.env`.
+6. Change the `BACKUP_BASE_NAME` in `local.env` if desired.  The default is `tfc-dump` which will create backup
+files named `tfc-dump-`_yyyy-mm-dd_`.tar.gz`.
+7. Obtain a 1Password Service Account token. See https://developer.1password.com/docs/service-accounts/get-started/ for details.
+8. Add the 1Password Service Account token to `local.env`.
+9. Build the Docker image:  docker build --tag dump-save:latest .
+9. Run the Docker image:  docker run --env-file=local.env dump-save:latest
```
Comment (briskt) [inline README.md]: Could you either use `1.` for all of the steps or correct this line to be `10.`?

File: dump-and-save.sh [hunk 7]
```diff
@@ -0,0 +1,29 @@
+#!/bin/sh
+
+# dump-and-save.sh - Dump Terraform Cloud workspaces and variables then save them to 1Password
+#
+# Dale Newby
+# SIL International
+# July 7, 2023
```
Comment (briskt) [inline dump-and-save.sh:7]: I'm curious about the purpose of these three lines (name, org, date). It seems redundant with git commit history.

File: local.env.dist [hunk 8]
```diff
@@ -0,0 +1,13 @@
+# ATLAS_TOKEN      - Terraform Cloud access token ($TFC_API_TOKEN)
+# onepasswordkey   - 1Password service account access key
+# ORGANIZATION     - Name of the Terraform Cloud organization to be backed up
+# VAULT_NAME       - 1Password vault name  "Apps Dev"
+# VAULT_ITEM       - Item within the 1Password vault  "Terraform Workspace Backups"
+# BACKUP_BASE_NAME - Base name of the backup file (${BACKUP_BASE_NAME}-yyyy-mm-dd.tar.gz)  "tfc-backup"
+
+ATLAS_TOKEN=
+onepasswordkey=
```
Comment (briskt) [inline local.env.dist]: Why is this lowercase and not word-separated? I don't see it referenced anywhere.

**Review by briskt — COMMENTED** [review]
While I'm a 1Password fan, it doesn't seem like a perfect fit for backups. Can you give me some background, please? (Here or in Slack)

File: Dockerfile (hunk 4 above)
Comment (dalenewby) [reply Dockerfile]: Agreed.  Its a habit to ensure that executable files remain executable even if they lose that attribute accidentally in a future version.  I'm not consistent in that I don't do this for packages I install from distribution repos.

File: Dockerfile (hunk 5 above)
Comment (dalenewby) [reply Dockerfile:15]: Agreed.  Still a habit.

File: README.md (hunk 6 above)
Comment (dalenewby) [reply README.md]: I knew I was going to mess that up when I started numbering them and moving them around.

File: dump-and-save.sh (hunk 7 above)
Comment (dalenewby) [reply dump-and-save.sh:7]: Habit.  It answers the question "Who wrote this and when?" when a user finds the file and doesn't know from whence it came.

File: local.env.dist (hunk 8 above)
Comment (dalenewby) [reply local.env.dist]: I changed the variable's name and forgot to update it here.

File: Dockerfile (hunk 3 above)
Comment (dalenewby) [reply Dockerfile:1]: I don't know that the Alpine version matters in this case.  3.18.2 is currently the latest version; 3.18 seemed like a safe and stable version to specify.

**Review by briskt — APPROVED** [review]

**Comment by dalenewby** [comment]
I've decide this is not the right way to perform the backups.

**Pull request closed without merging**